File: 2. Best Time to Buy and Sell Stock/solution.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        profit = 0
        buy_price = prices[0]
        for price in prices[1:]:
            profit = max(profit, price - buy_price)
            buy_price = min(price, buy_price)
        return profit

        

File: 2. Best Time to Buy and Sell Stock/solution.go
package main

import (
	"fmt"
	"math"
)

func main() {
	prices := []int{7, 1, 5, 3, 6, 4}
	result := maxProfit(prices)
	fmt.Println(result)
}

func maxProfit(prices []int) int {
	left, right, max := 0, 1, 0

	for right < len(prices) {
		profit := prices[right] - prices[left]
		if prices[left] < prices[right] {
			max = int(math.Max(float64(max), float64(profit)))
		} else {
			left = right
		}
		right += 1
	}
	return max
}